### DISCUSSION TITLE
[proposal] allow more compact 'using static' syntax

### DISCUSSION THREAD

**Discussion by glenn-slayden** [dotnet/csharplang#3719] [Language Ideas]
(for the purposes of demonstration):

	enum enumA { }
	struct structB { }
	namespace some_ns{ public class classC { } }

**C# 6.0** introduced the [`using static`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-static) directive:

	namespace other_ns
	{
		using static enumA;
		using static global::structB;
		using static nsX.classC;

		// etc...
	}

Proposing that a future version of **C#** should optionally allow the following more compact syntax:

	namespace other_ns
	{
		using static enumA, global::structB, nsX.classC;

		// etc...
	}

Note that this optional form would align `using static` syntax with existing syntax for (optionally) combining [local declarations](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#local-variable-declarations) for variables having the same `Type`.

**Comment by huoyaoyuan** [comment]
**I just don't understand why people are so persistent on "compact".**
It's always non-goal to make code shorter in file size, or lines. The goal should be making it easier to use in every way. *Theoretically, you can always type code character by character, but languages should not be designed for that.*

**Reading**: honestly, people seldom look at usings. Since there are separators(`.`, `::`) in the names being using-d, it will be harder to distinguish them if put in a single line, separated by comma.
**Tooling**: adding and removing `using`s line by line will obviously make git merge easier.
**Writing**: I haven't typed `using namespace` for a long time, thanks to VS's completion. For `using static`, currently there are no completion available, *but the compact syntax doesn't help*. If there is already a line of `using static`, I can just press `Ctrl+D` to duplicate the line, and change the class being using-d.

The key difference between `using` and expression bodies is that, **the latter constructs the majority of code**. Thus, making it more compact helps more.

**Comment by glenn-slayden** [comment]
***I just don't understand why people...***

The proposed feature is optional. You don't have to use it if you don't want to

***The goal should be making it easier to use in every way...***

 I agree, which is why being able to see and study more of the code on the screen at once without scrolling or other fiddling--which is what reducing line count does--is a considerable benefit.

***people seldom look at usings...***

You contradict yourself. If they are seldom looked at, then wouldn't you *want* to get them out of the way?

***...make git merge easier...***

Not all code is version controlled, but even so, see above, *viz.*, the proposed syntax is optional.

***I haven't typed `using namespace` for a long time...***

Non-sequitur. The proposal only concerns `using static`.

***The key difference between using and expression bodies...***

Well, another major difference is that `using` directives are usually at the top of the file. Having excess lines there is especially annoying, since you always have to page down past them every time a file is first opened.

**Comment by CyrusNajmabadi** [comment]
> You contradict yourself. If they are seldom looked at, then wouldn't you want to get them out of the way?

I think where i land is simply this:

i dont' think about `usings` and i'm surprised to find out when other people do.  I *think* (though would have to even check since i never look at them), that they're just always collapsed in my files.  And my IDE just adds or removes the usings so i don't have to care about tehm.

I don't want to invest in language features here because it's genuinely an area of a person's code that i effectively don't expect them to be editing.  And i don't see there being real value in changing the language here for things that users will not be touching the majority of the time.

**Comment by glenn-slayden** [comment]
> i (sic.) don't have to care about tehm (sic.)...

If you don't care about them, then this proposal shouldn't affect you, so I'm not sure why you have an opinion about it.

> area of ... code that i effectively don't expect them to be editing...

@CyrusNajmabadi `using` declarations are part of the language, regardless of your personal coding style or choice of IDE.

I'm brand new to this forum, so I'm not sure if you are an official guardian of **C#**. I'm also not sure about the militancy of discussion culture accepted here, so I certainly don't want to offend. Given those caveats, I humbly submit that it seems to me like you "expect" everyone to use, focus on, ignore, hide, or automate the various language features in a universally specific way.

Again, apologies if the goal of eliminating variation in the personal style and use C# (not sarcastic) is a widely acknowledged consensus of this forum.

**Comment by theunrepentantgeek** [comment]
> Well, another major difference is that using directives are usually at the top of the file. Having excess lines there is especially annoying, since you always have to page down past them every time a file is first opened.

This doesn't match with my experience. 

When I navigate to a definition, I end up at the definition, not at the top of the file.
When I navigate to an implementation, I end up at the declaration of the implementation, not at the top of a file.
When I trace into a method while debugging, the file opens to that method, not at the top of a file.

Same when I jump to a failing unit test, ask git to show me changes, review diffs before making a commit, reviewing a PR, and so on. 

> The proposed feature is optional. You don't have to use it if you don't want to

You're arguing against yourself here. If the feature isn't going to be widely used, why should the team invest the (large) amount of effort to design, develop, test, and maintain it?

New features need to be more than just _useful_ or _good ideas_ or _helpful_. They need to be _massively valuable_. In practice, this seems to break down as either "widely useful to many millions of C# developers" (e.g. expression bodied members) or "critically valuable to a specific segment" (e.g. `Span<T>`)

> I'm brand new to this forum, so I'm not sure if you are an official guardian of C#. 

Welcome!

C# doesn't have individual guardians - there is a language development group within Microsoft that stewards the language. The members of that group are all here in this C# repo, with differing levels of engagement. @CyrusNajmabadi is a part of that group.

No matter how you measure it, changes to C# are expensive - both initially, and in the long term, as the language is maintained. Every proposed idea for the language has to stack up in a dozen different ways; very few of them actually get built.

> I'm also not sure about the militancy of discussion culture accepted here, so I certainly don't want to offend

As a whole we're hard to offend - as long as you're genuinely engaging with the discussion, you'll find it a lot of fun here. 

That said, _ideas_ will be critiqued, questions asked, and assumptions challenged. Some people find that tough going.

**Comment by glenn-slayden** [comment]
@theunrepentantgeek  wrote:
> New features need to be more than just useful or good ideas or helpful. They need to be massively valuable.

Thanks for the clarifications and guidance. It sounds like my proposal will be a dead letter. However, since I feel it is a good faith proposal that doesn't feel patently frivolous to me (do you disagree?), rather than close the issue at this point, I'll take the opportunity to observe and learn how the (presumably tortuous?) consideration process proceeds to some official conclusion,

**Comment by CyrusNajmabadi** [comment]
> If you don't care about them, then this proposal shouldn't affect you, so I'm not sure why you have an opinion about it.

A few reasons I have an opinion on it:

First, I'm a member of the language design group. So it's literally part of my job description to determine which features make it into c# and which do not :-)

Second, I am a developer on the C#/VB ide. So changes to the language have to be designed and implemented by us. 

Third, I've been involved with, and using C# for more than 20 years now. So that had definitely informed my perspective on where I think the language should focus it's priorities. 

> I humbly submit that it seems to me like you "expect" everyone to use, focus on, ignore, hide, or automate the various language features in a universally specific way.

I don't expect that. I simply don't think (and haven't gotten evidence), that an investment here is warranted. :-)

> I'll take the opportunity to observe and learn how the (presumably tortuous?) 

Feel free to engage. This repo is about discussion :-)   I don't think there's anything torturous here. It's just a process of evaluation and deciding what provided enough value to go in the language.

**Comment by chrisoverzero** [comment]
> (presumably tortuous?)

> I don't think there's anything torturous here.

@CyrusNajmabadi In case that wasn’t a typo, glenn-slayden is saying the process is presumably twisting or convoluted. As opposed to saying that it is excruciating or that it causes suffering. 

Though, @glenn-slayden, in my experience, it’s neither. It’s only that such a mature and widely used language has a very high bar for what gets into it.

**Comment by glenn-slayden** [comment]
> @CyrusNajmabadi wrote:
> ...member of the language design group, ... developer on the C#/VB ide, ... using C# for more than 20 years...

Not sure if I was blundering in here or not, but I did not know any of this; In any case, I apologize for any offense.

> @chrisoverzero wrote:
> ...very high bar for what gets into it

Indeed, I'm starting to understand that.

<br>_p.s._:<br>
![tortuous](https://user-images.githubusercontent.com/5589855/88473470-b0a0be80-ced2-11ea-8abf-0ce6396cb471.png)

**Comment by CyrusNajmabadi** [comment]
No offense taken. But it is important to me that opinions are welcome here. Even if they differ from the positions you are in favor of :-)
